File: scripts/lua-obfuscator.js
#!/usr/bin/env node
/**
 * VexonAC Lua obfuscator — FiveM-safe, zero external dependencies.
 *
 * Techniques applied per file:
 *  1. Strip all comments (-- and --[[ ]])
 *  2. Encode every string literal as string.char(byte, byte, ...)
 *  3. Rename all local variables to random hex identifiers
 *  4. XOR-encode the processed source and emit a self-decoding loader
 *
 * Usage: node lua-obfuscator.js <input.lua> <output.lua>
 */

const fs   = require("fs");
const path = require("path");
const crypto = require("crypto");

// ─── helpers ────────────────────────────────────────────────────────────────

function randomIdent(len = 12) {
  return "_" + crypto.randomBytes(len).toString("hex");
}

function xorKey(len = 32) {
  return Array.from(crypto.randomBytes(len));
}

// ─── Step 1: strip comments ─────────────────────────────────────────────────

function stripComments(src) {
  // Remove long comments --[[ ... ]]
  src = src.replace(/--\[\[[\s\S]*?\]\]/g, "");
  // Remove short comments -- ... (not inside strings)
  const lines = src.split("\n");
  const out = [];
  for (const line of lines) {
    // Simple: remove anything after -- that isn't inside a string.
    // We walk char-by-char to be safe.
    let result = "";
    let inStr = false;
    let strChar = "";
    let i = 0;
    while (i < line.length) {
      const c = line[i];
      if (!inStr && (c === '"' || c === "'")) {
        inStr = true; strChar = c; result += c; i++;
        continue;
      }
      if (inStr && c === strChar && line[i - 1] !== "\\") {
        inStr = false; result += c; i++;
        continue;
      }
      if (!inStr && c === "-" && line[i + 1] === "-") break; // comment starts
      result += c; i++;
    }
    out.push(result);
  }
  return out.join("\n");
}

// ─── Step 2: encode string literals as string.char(...) ─────────────────────

function encodeStrings(src) {
  // Match single or double quoted strings (non-multiline), avoid long strings [[ ]]
  return src.replace(/"((?:[^"\\]|\\.)*)"|'((?:[^'\\]|\\.)*)'/g, (match, d, s) => {
    const raw = d !== undefined ? d : s;
    // Decode escape sequences to actual characters
    let decoded;
    try {
      // Use JSON.parse for double-quoted; handle single-quoted manually
      decoded = d !== undefined
        ? JSON.parse('"' + raw.replace(/\n/g, "\\n") + '"')
        : raw.replace(/\\n/g, "\n").replace(/\\t/g, "\t").replace(/\\'/g, "'").replace(/\\\\/g, "\\");
    } catch {
      return match; // leave unchanged if we can't parse
    }
    const bytes = [];
    for (let i = 0; i < decoded.length; i++) {
      bytes.push(decoded.charCodeAt(i));
    }
    if (bytes.length === 0) return '""';
    return "string.char(" + bytes.join(",") + ")";
  });
}

// ─── Step 3: rename local variables ─────────────────────────────────────────

function renameLocals(src) {
  const map = new Map();
  // Match `local varName` and `local varName,` and function params
  src = src.replace(/\blocal\s+([a-zA-Z_][a-zA-Z0-9_]*)\b/g, (match, name) => {
    if (!map.has(name)) map.set(name, randomIdent());
    return "local " + map.get(name);
  });
  // Replace all usages of renamed vars
  for (const [original, renamed] of map.entries()) {
    // Only replace word-boundary matches that aren't preceded by a dot (method calls)
    const re = new RegExp("(?<!\\.)\\b" + original + "\\b", "g");
    src = src.replace(re, renamed);
  }
  return src;
}

// ─── Step 4: XOR-encode the entire source and emit a loader ─────────────────

function xorWrap(src) {
  const key = xorKey(64);
  const bytes = Buffer.from(src, "utf8");
  const encoded = Buffer.alloc(bytes.length);
  for (let i = 0; i < bytes.length; i++) {
    encoded[i] = bytes[i] ^ key[i % key.length];
  }

  const encodedHex = Array.from(encoded).join(",");
  const keyHex     = key.join(",");

  // The loader is intentionally minimal and hard to follow
  const loaderVar   = randomIdent();
  const keyVar      = randomIdent();
  const dataVar     = randomIdent();
  const iVar        = randomIdent();
  const outVar      = randomIdent();
  const fnVar       = randomIdent();

  return [
    `local ${keyVar}={${keyHex}}`,
    `local ${dataVar}={${encodedHex}}`,
    `local ${outVar}={}`,
    `for ${iVar}=1,#${dataVar} do`,
    `  ${outVar}[${iVar}]=string.char(${dataVar}[${iVar}]~${keyVar}[((${iVar}-1)%#${keyVar})+1])`,
    `end`,
    `local ${loaderVar}=table.concat(${outVar})`,
    `local ${fnVar},${loaderVar}=load(${loaderVar})`,
    `if ${fnVar} then ${fnVar}() end`,
  ].join("\n");
}

// ─── Main ────────────────────────────────────────────────────────────────────

function obfuscate(src) {
  let out = src;
  out = stripComments(out);
  out = encodeStrings(out);
  out = renameLocals(out);
  out = xorWrap(out);
  return out;
}

const [,, input, output] = process.argv;
if (!input || !output) {
  console.error("Usage: node lua-obfuscator.js <input.lua> <output.lua>");
  process.exit(1);
}

const src = fs.readFileSync(input, "utf8");
const result = obfuscate(src);
fs.mkdirSync(path.dirname(output), { recursive: true });
fs.writeFileSync(output, result, "utf8");
console.log(`  obfuscated: ${path.basename(input)} → ${path.basename(output)}`);

File: scripts/lua-obfuscator.py
#!/usr/bin/env python3
"""
VexonAC Lua obfuscator — FiveM-safe, zero deps beyond stdlib.
Techniques:
  1. Strip -- and --[[ ]] comments
  2. Encode string literals as string.char(byte, ...)
  3. Rename local variables to random hex names
  4. XOR-encrypt full source + emit self-decoding loader
"""

import sys, re, os, secrets, json

# ── 1. strip comments ─────────────────────────────────────────────────────────

def strip_comments(src: str) -> str:
    # Long comments first
    src = re.sub(r'--\[\[.*?\]\]', '', src, flags=re.DOTALL)
    lines = []
    for line in src.split('\n'):
        result = []
        in_str, str_char, i = False, '', 0
        while i < len(line):
            c = line[i]
            if not in_str and c in ('"', "'"):
                in_str, str_char = True, c
                result.append(c); i += 1; continue
            if in_str and c == str_char and (i == 0 or line[i-1] != '\\'):
                in_str = False
                result.append(c); i += 1; continue
            if not in_str and c == '-' and i+1 < len(line) and line[i+1] == '-':
                break
            result.append(c); i += 1
        lines.append(''.join(result))
    return '\n'.join(lines)

# ── 2. encode string literals ─────────────────────────────────────────────────

def encode_strings(src: str) -> str:
    def replacer(m):
        raw = m.group(1) if m.group(1) is not None else m.group(2)
        try:
            # Decode Lua escape sequences
            decoded = raw.replace(r'\\', '\x00BACKSLASH\x00')
            decoded = decoded.replace(r'\n', '\n').replace(r'\t', '\t')
            decoded = decoded.replace(r"\'", "'").replace(r'\"', '"')
            decoded = decoded.replace('\x00BACKSLASH\x00', '\\')
        except Exception:
            return m.group(0)
        if not decoded:
            return '""'
        bytes_list = [str(ord(c)) for c in decoded]
        return 'string.char(' + ','.join(bytes_list) + ')'

    # Match double-quoted strings
    src = re.sub(r'"((?:[^"\\]|\\.)*)"', replacer, src)
    # Match single-quoted strings
    src = re.sub(r"'((?:[^'\\]|\\.)*)'", replacer, src)
    return src

# ── 3. rename locals ──────────────────────────────────────────────────────────

def rename_locals(src: str) -> str:
    name_map = {}

    def make_name():
        return '_' + secrets.token_hex(8)

    def map_name(original):
        if original not in name_map:
            name_map[original] = make_name()
        return name_map[original]

    # Find all `local <name>` declarations
    def replace_decl(m):
        return 'local ' + map_name(m.group(1))
    src = re.sub(r'\blocal\s+([a-zA-Z_][a-zA-Z0-9_]*)\b', replace_decl, src)

    # Replace usages (word boundary, not after a dot)
    for original, renamed in name_map.items():
        src = re.sub(r'(?<!\.)(?<![a-zA-Z0-9_])\b' + re.escape(original) + r'\b(?![a-zA-Z0-9_])', renamed, src)

    return src

# ── 4. XOR-wrap ───────────────────────────────────────────────────────────────

def xor_wrap(src: str) -> str:
    key = list(secrets.token_bytes(64))
    data = list(src.encode('utf-8'))
    encoded = [(b ^ key[i % len(key)]) for i, b in enumerate(data)]

    kv  = '_' + secrets.token_hex(6)
    dv  = '_' + secrets.token_hex(6)
    iv  = '_' + secrets.token_hex(6)
    ov  = '_' + secrets.token_hex(6)
    fv  = '_' + secrets.token_hex(6)
    lv  = '_' + secrets.token_hex(6)

    lines = [
        f'local {kv}={{{",".join(map(str,key))}}}',
        f'local {dv}={{{",".join(map(str,encoded))}}}',
        f'local {ov}={{}}',
        f'for {iv}=1,#{dv} do',
        f'  {ov}[{iv}]=string.char({dv}[{iv}]~{kv}[(({iv}-1)%#{kv})+1])',
        f'end',
        f'local {lv}=table.concat({ov})',
        f'local {fv},{lv}=load({lv})',
        f'if {fv} then {fv}() end',
    ]
    return '\n'.join(lines)

# ── main ──────────────────────────────────────────────────────────────────────

def obfuscate(src: str) -> str:
    src = strip_comments(src)
    src = encode_strings(src)
    src = rename_locals(src)
    src = xor_wrap(src)
    return src

if __name__ == '__main__':
    if len(sys.argv) != 3:
        print('Usage: python3 lua-obfuscator.py <input.lua> <output.lua>')
        sys.exit(1)
    inp, out = sys.argv[1], sys.argv[2]
    src = open(inp, 'r', encoding='utf-8').read()
    result = obfuscate(src)
    os.makedirs(os.path.dirname(out) or '.', exist_ok=True)
    open(out, 'w', encoding='utf-8').write(result)
    print(f'  obfuscated: {os.path.basename(inp)} → {os.path.basename(out)}')
